activerecord/test/cases/relation/where_test.rb

```diff
@@ -458,6 +458,13 @@ def test_where_on_association_with_select_relation
       assert_equal essays(:david_modest_proposal), essay
     end
 
+    def test_where_on_association_with_collection_polymorphic_relation
+      expected = Treasure.where(price_estimates: PriceEstimate.where(estimate_of_type: "Treasure"))
+      actual   = Treasure.where(price_estimates: PriceEstimate.all)
+
+      assert_equal expected.to_sql, actual.to_sql
```

Comment:
I fear this test might be a bit too dynamic, it could easily end up passing even if the actual feature is broken, as it rely a bit too much on higher level APIs.

Could we have a test similar to the one reported in the issue? One that actually perform the query and ensure only the record we want are returned?

Reply:
Now we have a Car and Treasure with the same id. Before this fix, it was returning emerald that in this case doesn't have a PriceEstimate and wrongly relating to honda Car just because both have the same id.

https://github.com/rails/rails/blob/a3e85f7c6f3e3922b3a93c8c8603e63465502b71/activerecord/test/cases/relation/where_test.rb#L461-L464